src/wazuh_modules/vulnerability_detector/wm_vuln_detector.c

```diff
@@ -5282,19 +5281,48 @@ void wm_vuldet_create_and_send_message(vu_task_ctx_t* task, task_status TASK_STA
     }
 }
 
-void wm_vuldet_on_demand_scan(vu_task_ctx_t* current_task, wm_vuldet_t* vuldet) {
+bool wm_vuldet_on_demand_scan(vu_task_ctx_t* current_task, wm_vuldet_t* vuldet) {
+    bool retry_agent = false;
+    char status_message[OS_SIZE_64] = {0};
+
+    if (!current_task || !vuldet) {
+        return false;
+    }
+
+    if (!current_task->agent_data) {
+        os_calloc(1, sizeof(scan_agent), current_task->agent_data);
+    } else {
+        time_t sleep_time = current_task->agent_last_scan_time + vuldet->retry_interval - time(NULL);
+        if (sleep_time > 0) {
+            mtdebug1(WM_VULNDETECTOR_LOGTAG, "Going to sleep %ld seconds before retrying pending agents", sleep_time);
+            sleep(sleep_time);
+        }
+    }
+
     wm_vuldet_create_and_send_message(current_task, WM_TASK_IN_PROGRESS, NULL);
-    if (wm_vuldet_collect_agent_info(current_task->agent_to_scan, vuldet)) {
-        mterror(WM_VULNDETECTOR_LOGTAG, VU_NO_AGENT_ERROR);
+    if (current_task->agent_last_scan_time == 0 && wm_vuldet_collect_agent_info(current_task->agent_to_scan, vuldet, current_task->agent_data)) {
+        mterror(WM_VULNDETECTOR_LOGTAG, VU_NO_AGENT_ERROR, current_task->agent_to_scan);
         wm_vuldet_create_and_send_message(current_task, WM_TASK_FAILED, NULL);
     } else {
-        if (wm_vuldet_run_scan(vuldet)) {
+        vuldet->agent_to_scan = current_task->agent_data;
+        if (wm_vuldet_check_agent_vulnerabilities(vuldet)) {
             wm_vuldet_create_and_send_message(current_task, WM_TASK_FAILED, NULL);
             mterror(WM_VULNDETECTOR_LOGTAG, VU_AG_CHECK_ERR);
         } else {
-            wm_vuldet_create_and_send_message(current_task, WM_TASK_DONE, NULL);
+            if (current_task->agent_data->pending_attempts > 0) {
+                retry_agent = true;
+                snprintf(status_message, OS_SIZE_64, " Agent software not ready. It will be tried again later.");
+                current_task->agent_last_scan_time = time(NULL);
+                wm_vuldet_create_and_send_message(current_task, WM_TASK_PENDING, status_message);
+            } else if (current_task->agent_data->pending_attempts < 0){
+                wm_vuldet_create_and_send_message(current_task, WM_TASK_FAILED, NULL);
+            } else {
+                wm_vuldet_create_and_send_message(current_task, WM_TASK_DONE, NULL);
+            }
         }
     }
+
+    return retry_agent;
 }
 
 void wm_vuldet_on_demand_feed_update(vu_task_ctx_t* current_task, wm_vuldet_t* vuldet) {
```

Comment:
I would unify this with wm_vuldet_feed_update_on_demand

Reply:
The name were unified